### ISSUE TITLE
C# 3.0.3-stable signals not always triggering with exported binaries

### ISSUE THREAD

**Issue by exts** [godotengine/godot#19563]
**Godot version:**
3.0.3 Stable

**OS/device including version:**
Windows 10, Linux (Ubuntu 18) and OSX (macOS Sierra 10.12.6 - friend tested)

**Issue description:**
Problem is when you export a C# project, signals sometimes will not trigger. When I say sometimes I mean the debug version locally works flawlessly. On other pc's the debug has issue where my WeaponDischarged signal doesn't trigger. Release version on other pc's it'll work once or twice then fail to trigger the signals after trying to restart the game again.

I could've sworn this issue was fixed, but it seems like it's back

```
ERROR: emit_signal: Can't emit non-existing signal "WeaponDischarged".
    At: core/object.cpp:1139.
```

**Steps to reproduce:**
1. Create a game with signal usage. 
2. Export game
3. Watch console spit out errors about signal

**Minimal reproduction project:**
[SpaceShooter.zip](https://github.com/godotengine/godot/files/2102785/SpaceShooter.zip)
https://github.com/exts/godotsharp-shooter

Binaries: https://g4mr.itch.io/testing-godot

(Controls: WASD and SPACE, space to shoot)

**Comment by akien-mga** [comment]
CC @neikeq. Setting milestone 3.1 for now until we identify if it's reproducible in the master branch or a 3.0-only issue.

**Comment by heikytr** [comment]
I could reproduce this using the last commit 4da621603eaf4b85dbaacb64a010ec2e62574ec5 from master using this example:
```Csharp
public class TestClass : Node
{
    [Signal]
    delegate void TestSignal(string message);

    public override void _Ready()
    {
        EmitSignal("TestSignal", "Hello World!");
    }
}
```
```Csharp
public class ReceiveSignal : Label
{
    public override void _Ready()
    {
        GetParent().Connect("TestSignal", this, "PrintMessage");
    }

    void PrintMessage(string message)
    {
        GD.Print(message);
        SetText(message);
    }
}
```
Although 7 out of 10 tries it will just "work", sometimes the signal will not connect giving me this error:

> ERROR: In Object of type 'Node': Attempt to connect nonexistent signal 'TestSignal' to method 'Label.PrintMessage'
>    At: core\object.cpp:1404
> ERROR: Can't emit non-existing signal "TestSignal".
>    At: core\object.cpp:1139

**Comment by malbach** [comment]
Hi everyone,

Can also confirm the bug using master from [5c5aafa](https://github.com/godotengine/godot/commit/5c5aafabecfeaa056cae4c3a7fbaa1c52f147930)

But from the OP's example project, I modified the GameScene script from
```C#
// connect the ship signal to a method inside our game scene, this would allow us to spawn bullets
// into the scene after we press spacebar. Allowing us to keep our code as separated as possible. Also
// allowing us to have bullets interact with other things like enemies in our case.
_ship = (Ship) GetNode("GameCanvas/Ship");
_ship.Connect("WeaponDischarged", this, "SpawnBullet");
```
to
```C#
// connect the ship signal to a method inside our game scene, this would allow us to spawn bullets
// into the scene after we press spacebar. Allowing us to keep our code as separated as possible. Also
// allowing us to have bullets interact with other things like enemies in our case.
_ship = (Ship) GetNode("GameCanvas/Ship");
_ship.AddUserSignal("WeaponDischarged");  // <<< add signal at setup
_ship.Connect("WeaponDischarged", this, "SpawnBullet");
```
and removed the signal declaration in Ship.cs
```C#
    public class Ship : Area2D
    {
        //[Signal]
        //public delegate void WeaponDischarged();
```
and it is working as it should, even exported in release.
I'm jaust having problems with the IsActionJustPressed in Ship.cs that seem to not be triggered, so I replaced it with
```C#
    Input.IsKeyPressed((int) KeyList.Space)
```
for the tests.

**Comment by akien-mga** [comment]
CC @paulloz

**Comment by paulloz** [comment]
Thanks for the ping 😀, I don't have enough free time these days to check issues without being notified 😅.  
I'll check this tomorrow.

**Comment by exts** [comment]
Oh this is interesting @malbach. Updated my code, removed all signal annotations and just called AddUserSignal and my linux build works normally now O_o.

```csharp
_ship = (Ship) GetNode("GameCanvas/Ship");
_ship.AddUserSignal("WeaponDischarged");
_ship.Connect("WeaponDischarged", this, "SpawnBullet");
```

```csharp
var enemy = (Enemy) _enemyObject.Instance();
enemy.Position = new Vector2(_formations.XPosition, _formations.Positions[spawn]);
enemy.AddUserSignal("Destroyed");
enemy.Connect("Destroyed", this, "EnemyDestroyed");
```

Even emitting values to the signal works. 

I uploaded a linux build here https://g4mr.itch.io/testing-godot called `linux64-6202018.zip`

**Comment by PJB3005** [comment]
FWIW, this is probably related to exports not being a tools build. There's probably some code not getting ran to handle delegate signals in those.

**Comment by exts** [comment]
@paulloz I think I figured out the issue, but unsure how to fix it. `void CSharpLanguage::reload_assemblies_if_needed` method is tied to the tools ifdef which is the main code that updates the signals map which is then loaded from godot via `void CSharpScript::get_script_signal_list` [right](https://github.com/godotengine/godot/blob/255d6470eec957b2d5010bd494f357aff1f7e8c6/core/object.cpp#L1325)? If someone can figure out how to sort that then hypothetically it should work outside of the editor.

Maybe @neikeq could take a look I know you're very busy with personal life stuff. My C++ is pretty dated.

**Comment by exts** [comment]
So I did test one last time just to confirm my findings. 
![godot_signal_issue](https://user-images.githubusercontent.com/16387107/42127114-b68caf80-7c58-11e8-8cc7-3dfb05f7ae96.png)

I basically just updated the `_get_signal` to post a few `ERR_PRINTS` just to see me stepping into each if statement [here](https://github.com/godotengine/godot/blob/master/modules/mono/csharp_script.cpp#L1615)  (no idea how to debug the editor/export templates) and when I built the editor off of 8bf5b502223d48ee1b7b88ab8613be7bacbb63d2 (since more recent commits were bugged) the editor showed me that's how the signals are being called.

Well in the debug export template only the instanced signal annotation was loaded via `ScriptInstance *CSharpScript::instance_create(Object *p_this)` which makes sense and the ship annotation not being loaded also make sense because it's only called in the reload assemblies if needed method so outside of the editor at runtime that code never gets executed. Hope that helps.

Here's a screenshot of the editor when I ran the test.
![image](https://user-images.githubusercontent.com/16387107/42127182-91d654ba-7c59-11e8-9c05-cb8d94121334.png)
